hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java

```diff
@@ -673,9 +670,10 @@ protected void getAllObjectInternal(List<Data> keys, List<Object> resultingKeyVa
      * Probably this can be optimized in the future by making use of an PartitionIterating operation.
      */
     protected void putAllInternal(Map<?, ?> map) {
-        int partitionCount = partitionService.getPartitionCount();
-
         try {
+            int partitionCount = partitionService.getPartitionCount();
+            int initialSize = map.size() / partitionCount;
```

Comment:
This is an educated guess to create the `MapEntires` with a proper size. If we don't do this we will suffer badly from constant resizing. The `MapEntires` for each partitionId are anyway created lazy, so the worst case would be a very unbalanced input map, where this `initialSize` is not sufficient and we'll get a bunch of resizes. But at least we won't waste memory on partitions, which are not used at all.